### ISSUE TITLE
TypeHintDeclarationSniff does not consider array of arrays

### ISSUE THREAD

**Issue by marcospassos** [slevomat/coding-standard#553]
The inspection reports the absence of type hint in the return and parameter of a method if the type is an array of arrays (`array[]`).

```php
/**
 * @param array[] $arrays
 *
 * @return array[]
 */
public function getAdditionTests(array $arrays) : array
{
    return [
        '1 + 2 = 3' => [
            'augend' => 1,
            'addend' => 2,
            'result' => 3,
        ],
    ];
}
```
The above snippet produces the following non-applicable warning:
```
@param annotation of getAdditionTests() does not specify type hint for items of its traversable parameter $arrays.
@return annotation of getAdditionTests() does not have @return annotation for its traversable return value.
```

**Comment by kukulich** [comment]
You should use `mixed[][]` or something like this.

**Comment by lookyman** [comment]
> You should use `mixed[][]` or something like this.

Why?

**Comment by marcospassos** [comment]
@kukulich you could be more open to the bug report, as we also spend time trying to report an issue not only because it affects us, but also to help make the library more robust. [This is the second issue you closed providing no feedback about the problem reported.](https://github.com/slevomat/coding-standard/issues/551) _Closing an issue does not fix it._

The annotation `@array[]` is a perfect valid PHPDoc annotation, that implies an array of arrays of mixed values. Why should I replace such a simple notation for something much less clear like `mixed[][]`?

**Comment by marcospassos** [comment]
But I'm doing exactly it: array is the type of the items. What is wrong with having an array of tuples, for instance?

```php
@param array[] $arrays
function foo() : array {
    return [[1, 2], [3, 4]];
}
```

Besides this, if `mixed[][]` is valid, it implies that`array[]` should work as well, as it means an array of arrays containg mixed values.

**Comment by kukulich** [comment]
No, `array` is not valid in this sniff, deal with it :)

**Comment by lookyman** [comment]
@kukulich I will deal with it (has been doing that since I started using this cs), but I still don't understand your reasoning.
